ndk-build/src/manifest.rs

```diff
@@ -1,258 +1,345 @@
+    #[serde(rename(serialize = "android:label"))]
+    pub label: Option<String>,
+    #[serde(rename(serialize = "android:launchMode"))]
+    pub launch_mode: Option<String>,
+    #[serde(rename(serialize = "android:name"))]
+    #[serde(default = "default_activity_name")]
+    pub name: Option<String>,
+    #[serde(rename(serialize = "android:screenOrientation"))]
+    pub orientation: Option<String>,
+
+    #[serde(rename(serialize = "meta-data", deserialize = "metadata"))]
+    pub meta_datas: Option<Vec<MetaData>>,
+    /// If no `MAIN` action exists in any intent filter, a default `MAIN` filter is serialized.
+    #[serde(serialize_with = "serialize_intents")]
+    #[serde(rename(serialize = "intent-filter", deserialize = "intent_filter"))]
+    pub intent_filters: Option<Vec<IntentFilter>>,
 }
 
-impl ApplicationMetadata {
-    pub fn to_string(&self) -> String {
-        format!(
-            r#"<meta-data android:name="{}" android:value="{}"/>"#,
-            self.name, self.value
-        )
+impl Default for Activity {
+    fn default() -> Self {
+        Self {
+            config_changes: default_config_changes(),
+            label: None,
+            launch_mode: None,
+            name: default_activity_name(),
+            orientation: None,
+            meta_datas: None,
+            intent_filters: None,
+        }
     }
 }
 
-#[derive(Debug)]
-pub struct ActivityMetadata {
-    pub name: String,
-    pub value: String,
-}
+fn serialize_intents<S>(
+    intent_filters: &Option<Vec<IntentFilter>>,
+    serializer: S,
+) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    let main_intent_filter = IntentFilter {
+        actions: vec!["android.intent.action.MAIN".to_string()],
+        categories: Some(vec!["android.intent.category.LAUNCHER".to_string()]),
+        data: None,
+    };
 
-impl ActivityMetadata {
-    pub fn to_string(&self) -> String {
-        format!(
-            r#"<meta-data android:name="{}" android:value="{}"/>"#,
-            self.name, self.value
-        )
+    // Check if `intent_filters` contains a `MAIN` action if not, add a default filter.
+    match intent_filters {
+        Some(filters) => {
+            if filters
+                .iter()
+                .all(|i| i.actions.iter().all(|f| f != "android.intent.action.MAIN"))
+            {
+                let mut filters = filters.clone();
+                filters.push(main_intent_filter);
+                return serializer.serialize_some(&filters);
+            }
+            serializer.serialize_some(filters)
+        }
+        None => serializer.serialize_some(&vec![main_intent_filter]),
     }
 }
 
-#[derive(Debug)]
-pub struct Permission {
-    pub name: String,
-    pub max_sdk_version: Option<u32>,
+/// See https://developer.android.com/guide/topics/manifest/intent-filter-element
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+pub struct IntentFilter {
+    /// Serialize as struct for proper xml formatting
```

Comment:
```suggestion
    /// Serialize wrapped in `<action android:name="..." />`
```
Or something like that, making it clear what "proper xml" is supposed to mean (text nodes in xml are valid, just not what Android expects). Same below.

Reply:
Rapped :joy: